src/slam_toolbox_common.cpp

```diff
@@ -44,15 +44,116 @@ SlamToolbox::SlamToolbox(rclcpp::NodeOptions options)
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+  first_measurement_ = true;
+  process_near_pose_ = nullptr;
+  configure();
+  loadPoseGraphByParams();
+  return CallbackReturn::SUCCESS;
+}
+
+/*****************************************************************************/
+CallbackReturn SlamToolbox::on_activate(const rclcpp_lifecycle::State &)
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+  sst_->on_activate();
+  sstm_->on_activate();
+  pose_pub_->on_activate();
+  closure_assistant_->on_activate();
+  scan_filter_ =
+    std::make_unique<tf2_ros::MessageFilter<sensor_msgs::msg::LaserScan>>(
+    *scan_filter_sub_, *tf_, odom_frame_, scan_queue_size_,
+    get_node_logging_interface(), get_node_clock_interface(),
+    tf2::durationFromSec(transform_timeout_.seconds()));
+  scan_filter_->registerCallback(
+    std::bind(&SlamToolbox::laserCallback, this, std::placeholders::_1));
+  return CallbackReturn::SUCCESS;
+}
+
+/*****************************************************************************/
+CallbackReturn SlamToolbox::on_deactivate(const rclcpp_lifecycle::State &)
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+  sst_->on_deactivate();
+  sstm_->on_deactivate();
+  pose_pub_->on_deactivate();
+  closure_assistant_->on_deactivate();
+  scan_filter_.reset();
+  return CallbackReturn::SUCCESS;
+}
+
+/*****************************************************************************/
+CallbackReturn SlamToolbox::on_cleanup(const rclcpp_lifecycle::State &)
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+  for (int i = 0; i != threads_.size(); i++) {
+    threads_[i]->join();
+    threads_[i].reset();
+  }
+  threads_.clear();
+
+  // delete in reverse order of initialization
+  closure_assistant_.reset();
+  map_saver_.reset();
+  scan_holder_.reset();
+  pose_helper_.reset();
+  laser_assistant_.reset();
+  solver_.reset();
+
+  scan_filter_sub_.reset();
+  ssDesserialize_.reset();
+  ssSerialize_.reset();
+  ssPauseMeasurements_.reset();
+  ssMap_.reset();
+  sstm_.reset();
+  sst_.reset();
+  pose_pub_.reset();
+
+  tfB_.reset();
+  tfL_.reset();
+  tf_.reset();
+
+  lasers_.clear();
+  dataset_.reset();
+  smapper_->getMapper()->Reset();
+  smapper_.reset();
+  return CallbackReturn::SUCCESS;
+}
+
+/*****************************************************************************/
+CallbackReturn
+SlamToolbox::on_shutdown(const rclcpp_lifecycle::State & state)
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  if (state.id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
+    on_deactivate(rclcpp_lifecycle::State());
```

Comment:
Remove. Shutdown cannot be called if not already in inactive state.

Reply:
OK.

I can manually transition to shutdown even if it is active. Should I return `CallbackReturn::FAILURE` just in case?

```console
$ ros2 lifecycle get /slam_toolbox 
active [3]
$ ros2 lifecycle list /slam_toolbox 
- deactivate [4]
	Start: active
	Goal: deactivating
- shutdown [7]
	Start: active
	Goal: shuttingdown
```

Reply:
Huh, I actually didn't know that.

Reply:
Oh, okay.

I looked at ros2 design and this transition (active -> shutdown) was also illustrated (<https://design.ros2.org/articles/node_lifecycle.html>).  
So I have left it as an expected transition. Please let me know if it needs to be updated again.